src/Glimpse.Common/Broker/DefaultMessageConverter.cs

```diff
@@ -18,36 +33,88 @@ public DefaultMessageConverter(JsonSerializer jsonSerializer, IContextData<Messa
         }
 
         public IMessage ConvertMessage(object payload)
-        { 
-            var message = new Message();
-            message.Id = Guid.NewGuid();
-            message.Type = payload.GetType().FullName;
-            message.Payload = _jsonSerializer.Serialize(payload);
-            message.Context = _contextData.Value;
+        {
+            var message = new Message
+            {
+                Id = Guid.NewGuid(),
+                Types = GetTypes(payload),
+                Payload = _jsonSerializer.Serialize(payload),
+                Context = _contextData.Value,
+                Indices = GetIndices(payload)
+            };
 
-            ProcessIndices(payload, message);
-            ProcessTags(payload, message);
+            message.Payload = _jsonSerializer.Serialize(message);
 
             return message;
-        } 
+        }
 
-        private void ProcessIndices(object payload, Message message)
+        private static IEnumerable<string> GetTypes(object payload)
         {
-            var indicesPayload = payload as IMessageIndices;
-            if (indicesPayload?.Indices != null)
+            var typeInfo = payload.GetType().GetTypeInfo();
+
+            return typeInfo.BaseTypes(includeSelf: true)
+                .Concat(typeInfo.ImplementedInterfaces)
+                .Except(_exclusions)
+                .Select(t => t.KebabCase());
+        }
+
+        private static IReadOnlyDictionary<string, object> GetIndices(object payload)
```

Comment:
I wonder if its worth pulling out the indices and type stuff into separate providers provider? It would mean the complexity of the class is reduced a bit and if someone wanted to augment the behaviour they could. I'm not super passionate about it, but food for thought.